src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,400 @@
+                {
+                    return true;
+                }
+            }
+
+        Failure:
+            address = default;
+            prefixLength = default;
+            return false;
+        }
+
+        private static bool Validate(IPAddress baseAddress, int prefixLength, bool throwOnFailure)
+        {
+            int maxPrefixLength = baseAddress.AddressFamily == AddressFamily.InterNetwork ? 32 : 128;
+            if (prefixLength < 0 || prefixLength > maxPrefixLength)
+            {
+                if (throwOnFailure)
+                {
+                    ThrowArgumentOutOfRangeException();
+                }
+
+                return false;
+            }
+
+            if (HasNonZeroBitsAfterNetworkPrefix(baseAddress, prefixLength))
+            {
+                if (throwOnFailure)
+                {
+                    ThrowInvalidBaseAddressException();
+                }
+
+                return false;
+            }
+
+            return true;
+
+            [DoesNotReturn]
+            static void ThrowArgumentOutOfRangeException() => throw new ArgumentOutOfRangeException(nameof(prefixLength));
+
+            [DoesNotReturn]
+            static void ThrowInvalidBaseAddressException() => throw new ArgumentException(SR.net_bad_ip_network_invalid_baseaddress, nameof(baseAddress));
+        }
+
+        private static bool HasNonZeroBitsAfterNetworkPrefix(IPAddress baseAddress, int prefixLength)
+        {
+            if (baseAddress.AddressFamily == AddressFamily.InterNetwork)
+            {
+                uint mask = (uint)((long)uint.MaxValue << 32 - prefixLength);
+                if (BitConverter.IsLittleEndian)
+                {
+                    mask = BinaryPrimitives.ReverseEndianness(mask);
+                }
+
+                return (baseAddress.PrivateAddress & mask) != baseAddress.PrivateAddress;
+            }
+            else
+            {
+                Unsafe.SkipInit(out UInt128 value);
+                baseAddress.TryWriteBytes(MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref value, 1)), out _);
+                if (prefixLength == 0)
+                {
+                    return value != UInt128.Zero;
+                }
+
+                UInt128 mask = UInt128.MaxValue << 128 - prefixLength;
+                if (BitConverter.IsLittleEndian)
+                {
+                    mask = BinaryPrimitives.ReverseEndianness(mask);
+                }
+
+                return (value & mask) != value;
+            }
+        }
+
+        /// <summary>
+        /// Converts the instance to a string containing the <see cref="IPNetwork"/>'s CIDR notation.
+        /// </summary>
+        /// <returns>The <see cref="string"/> containing the <see cref="IPNetwork"/>'s CIDR notation.</returns>
+        public override string ToString()
+        {
+            return $"{BaseAddress}/{PrefixLength}";
```

Comment:
Not big fan of making `default(IPNetwork)` equivalent to another arbitrary state. Encountering such instances is a sign of a programming error in most of the cases. Would it be very unconventional if `ToString()` returned something like `<uninitialized>` instead?

Reply:
> IPAddress.Any

Note that it could be also `IPAddress.IPv6Any`.

Reply:
> Not big fan of making default(IPNetwork) equivalent to another arbitrary state.

I think we need to make `default(IPNetwork)` valid.  If you were forced to add a `new IPNetwork()` constructor, what would its behavior be?  What would an `IPNetwork.Empty` do?  Etc.  Whatever that is, I think we should make `default(IPNetwork)` behave equivalently.  Having the default value of a struct randomly fail is an anti-pattern.  We typically only do it on our structs not meant to be used directly, e.g. helper types used by a compiler or other tools.

Reply:
One can argue that `default(IPNetwork)` is an empty network that doesn't contain any `IPAddress`, however note that it's not possible to define such an instance with an IP/prefix-length pair. 

> Having the default value of a struct randomly fail is an anti-pattern.

I have got rid of the exceptions, so the current PR behavior of `default(IPNetwork)` is what I just described above ("empty network"). The methods doen't fail anymore, we just don't pretend that there is a CIDR pair when none was defined.

Personally I find this to be a good compromise, but if you think it is not acceptable, and we should assign an `IPAddress` to the `BaseAddress == null` case, I don't think it would be a consistent behavior to special-case it only only in method implementations, we should rather do something like:

```C#
public readonly struct IPNetwork
{
    private IPAddress _address;

    public IPAddress BaseAddress => _address ?? IPAddress.Any;
}
```

but I personally find this worse because `IPAddress.Any` is an arbitrary choice.

Reply:
default(IPNetwork) is all zeros.  Imagine IPAddress were a struct rather than a class.  What would its all zero's value be?  That's IPAddress.Any.  Hence why I was suggesting it behave like IPAddress.Any + PrefixLength==0.

From a quick look at the latest code, I think that would primarily just mean making Contains return true rather than false when BaseAddress is null, just as it returns true if PrefixLength == 0.

Also, I do think you should make `BaseAddress` return something if _address is null.  Otherwise, the nullable annotations are wrong: it's advertising it will always return something non-nullable, but it might return null.

Reply:
> That's IPAddress.Any

Or `IPAddress.IPv6Any`, depends on the implementation. I don't think this case is similar to eg. `default(Memory<T>)`. Using `default(IPNetwork)` is a programming bug in 99% of the cases. As a user, I would expect a sw library to help detecting it and "fail fast" instead of hiding it with some arbitrary "correct" behavior. But based on your feedback, I assume this is not our preference with BCL structs. I wish we had design guidelines covering such details.